In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.simplefilter("ignore")

In [2]:
#Read the data set 
data = pd.read_excel(r"/home/dell/Downloads/iris.xls")
data

SL   SW   PL   PW  Classification
0    5.1  3.5  1.4  0.2     Iris-setosa
1    4.9  3.0  1.4  0.2     Iris-setosa
2    NaN  3.2  1.3  0.2     Iris-setosa
3    4.6  3.1  1.5  0.2     Iris-setosa
4    5.0  3.6  1.4  0.2     Iris-setosa
..   ...  ...  ...  ...             ...
145  6.7  3.0  5.2  2.3  Iris-virginica
146  6.3  2.5  5.0  1.9  Iris-virginica
147  6.5  3.0  NaN  2.0  Iris-virginica
148  6.2  3.4  5.4  2.3  Iris-virginica
149  5.9  3.0  5.1  1.8  Iris-virginica

[150 rows x 5 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   SL              143 non-null    float64
 1   SW              144 non-null    float64
 2   PL              144 non-null    float64
 3   PW              150 non-null    float64
 4   Classification  150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [4]:
data.columns

Index(['SL', 'SW', 'PL', 'PW', 'Classification'], dtype='object')

In [5]:
data.shape

(150, 5)

In [6]:
#Label Encoding
from sklearn.preprocessing import LabelEncoder

In [7]:
le=LabelEncoder()
data['Classification']=le.fit_transform(data['Classification'])
data['Classification'].value_counts()

0    50
1    50
2    50
Name: Classification, dtype: int64

In [8]:
le.classes_

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

In [9]:
#checking for null values
data.isnull().sum()

SL                7
SW                6
PL                6
PW                0
Classification    0
dtype: int64

In [10]:
#filling the missing values by using their column mean
data = data.where(pd.notna(data), data.mean(), axis="columns")  
data

SL   SW       PL   PW  Classification
0    5.100000  3.5  1.40000  0.2               0
1    4.900000  3.0  1.40000  0.2               0
2    5.855944  3.2  1.30000  0.2               0
3    4.600000  3.1  1.50000  0.2               0
4    5.000000  3.6  1.40000  0.2               0
..        ...  ...      ...  ...             ...
145  6.700000  3.0  5.20000  2.3               2
146  6.300000  2.5  5.00000  1.9               2
147  6.500000  3.0  3.75625  2.0               2
148  6.200000  3.4  5.40000  2.3               2
149  5.900000  3.0  5.10000  1.8               2

[150 rows x 5 columns]

In [11]:
# Dataset divided into 2
x=data.iloc[:,:4]
y=data.iloc[:,4]   

In [12]:
x

SL   SW       PL   PW
0    5.100000  3.5  1.40000  0.2
1    4.900000  3.0  1.40000  0.2
2    5.855944  3.2  1.30000  0.2
3    4.600000  3.1  1.50000  0.2
4    5.000000  3.6  1.40000  0.2
..        ...  ...      ...  ...
145  6.700000  3.0  5.20000  2.3
146  6.300000  2.5  5.00000  1.9
147  6.500000  3.0  3.75625  2.0
148  6.200000  3.4  5.40000  2.3
149  5.900000  3.0  5.10000  1.8

[150 rows x 4 columns]

In [13]:
y

0      0
1      0
2      0
3      0
4      0
      ..
145    2
146    2
147    2
148    2
149    2
Name: Classification, Length: 150, dtype: int64

In [14]:
 #Split the Data Into Train and Test Datasets
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=0)

In [15]:
x_train.shape

(112, 4)

In [16]:
x_test.shape

(38, 4)

In [17]:
y_train.shape

(112,)

In [18]:
y_test.shape

(38,)

# Logistic Regression

In [19]:
#Create the Model (Classification)
from sklearn.linear_model import LogisticRegression
model=LogisticRegression()

In [20]:
model.fit(x_train,y_train)

LogisticRegression()

In [21]:
#predict the model
y_pred=model.predict(x_test)

In [23]:
y_pred     # 0= Iris-setosa
           # 1= Iris-versicolor
           # 2= Iris-virginica

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 2, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0, 2, 1, 0, 2, 2, 1, 0, 2])

In [24]:
#checking the accuracy of the model
from sklearn.metrics import accuracy_score,confusion_matrix,precision_score, recall_score

In [25]:
confusion_matrix(y_test,y_pred)

array([[13,  0,  0],
       [ 0, 15,  1],
       [ 0,  0,  9]])

In [26]:
accuracy=accuracy_score(y_test,y_pred)*100
print("Accuracy of the model is {:.2f}".format(accuracy))

Accuracy of the model is 97.37


In [27]:
### calculating the probabilities 

In [28]:
y_pred_prob = model.predict_proba(x_test)

In [29]:
y_pred_prob

array([[1.18276206e-04, 5.32530571e-02, 9.46628667e-01],
       [1.96232520e-02, 9.40314824e-01, 4.00619239e-02],
       [9.90256374e-01, 9.74342541e-03, 2.01089137e-07],
       [3.94599988e-06, 3.18286530e-02, 9.68167401e-01],
       [9.80065543e-01, 1.99338910e-02, 5.66377198e-07],
       [4.31188945e-06, 8.36331485e-03, 9.91632373e-01],
       [9.86113462e-01, 1.38861807e-02, 3.57535639e-07],
       [3.89398871e-03, 6.25438452e-01, 3.70667559e-01],
       [2.19822201e-03, 6.36004290e-01, 3.61797488e-01],
       [2.79829513e-02, 8.90382972e-01, 8.16340769e-02],
       [4.23550161e-04, 2.83889369e-01, 7.15687081e-01],
       [1.79721123e-02, 7.11297867e-01, 2.70730020e-01],
       [8.60779074e-03, 7.92170879e-01, 1.99221330e-01],
       [3.96961075e-03, 6.56646881e-01, 3.39383509e-01],
       [6.96910633e-03, 6.87754018e-01, 3.05276876e-01],
       [9.75715560e-01, 2.42838969e-02, 5.43283389e-07],
       [9.99935699e-03, 7.03076321e-01, 2.86924322e-01],
       [2.24333146e-02, 8.54061

In [30]:
y_pred = (y_pred_prob[:,1] > 0.37).astype(int)

In [55]:
accuracy_score(y_test,y_pred)

0.34210526315789475

# Random Forest Regression 

In [36]:
# Fitting Random Forest Regression to the dataset
# import the regressor
from sklearn.ensemble import RandomForestRegressor
  
 # create regressor object
regressor = RandomForestRegressor(n_estimators = 100, random_state = 0)
  
# fit the regressor with x and y data
regressor.fit(x, y)  

RandomForestRegressor(random_state=0)

In [39]:
from sklearn import metrics 
print()
 
# using metrics module for accuracy calculation
print("ACCURACY OF THE MODEL: ", metrics.accuracy_score(y_test, y_pred))


ACCURACY OF THE MODEL:  0.9736842105263158


# kNN classifier

In [40]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=13)

In [41]:
lst_acc = []
for i in range(5,16,2):
    knn = KNeighborsClassifier(n_neighbors=i)
    knn.fit(x_train,y_train)
    y_pred = knn.predict(x_test)
    acc = accuracy_score(y_test,y_pred)
    lst_acc.append(acc)

In [42]:
lst_acc

[0.9736842105263158,
 0.9736842105263158,
 0.9736842105263158,
 0.9736842105263158,
 0.9736842105263158,
 0.9736842105263158]

In [43]:
from sklearn.model_selection import cross_validate
cvs = cross_validate(knn,x,y,cv=10)

In [44]:
cvs

{'fit_time': array([0.00424886, 0.00384402, 0.0034008 , 0.00355959, 0.00356698,
        0.00312233, 0.00794935, 0.00439405, 0.00353241, 0.00383472]),
 'score_time': array([0.00506449, 0.00479555, 0.00429535, 0.00465727, 0.00583196,
        0.00432563, 0.0055809 , 0.00559878, 0.00866437, 0.00473094]),
 'test_score': array([1.        , 0.86666667, 1.        , 1.        , 1.        ,
        0.86666667, 0.93333333, 1.        , 0.93333333, 0.93333333])}

In [41]:
cvs['test_score'].mean()

0.9533333333333334

In [45]:
cvs = cross_validate(model,x,y,cv=10)

In [46]:
cvs

{'fit_time': array([0.08264661, 0.08617067, 0.09038305, 0.05576253, 0.04976153,
        0.05088115, 0.05463767, 0.05367112, 0.0557332 , 0.05559969]),
 'score_time': array([0.00239921, 0.00671244, 0.00655723, 0.00225115, 0.0023551 ,
        0.00225329, 0.00266862, 0.00244308, 0.00235629, 0.00242734]),
 'test_score': array([1.        , 0.86666667, 1.        , 0.93333333, 0.93333333,
        0.86666667, 0.86666667, 1.        , 0.93333333, 0.86666667])}

In [47]:
cvs['test_score'].mean()

0.9266666666666667

###Conclusion: Both logistic regression and random forest  classification models shares the same and highest accuracy value 97.37 and 0.9736842105263158.So we can conclude that they are accurate models for the given dataset.